table/block_based/filter_policy.cc

```diff
@@ -82,21 +82,32 @@ class XXPH3FilterBitsBuilder : public BuiltinFilterBitsBuilder {
     // requirements.
     if (hash_entries_info_.entries.empty() ||
         hash != hash_entries_info_.entries.back()) {
-      if (detect_filter_construct_corruption_) {
-        hash_entries_info_.xor_checksum ^= hash;
-      }
-      hash_entries_info_.entries.push_back(hash);
-      if (cache_res_mgr_ &&
-          // Traditional rounding to whole bucket size
-          ((hash_entries_info_.entries.size() %
-            kUint64tHashEntryCacheResBucketSize) ==
-           kUint64tHashEntryCacheResBucketSize / 2)) {
-        hash_entries_info_.cache_res_bucket_handles.emplace_back(nullptr);
-        Status s = cache_res_mgr_->MakeCacheReservation(
-            kUint64tHashEntryCacheResBucketSize * sizeof(hash),
-            &hash_entries_info_.cache_res_bucket_handles.back());
-        s.PermitUncheckedError();
-      }
+      AddHash(hash);
+    }
+  }
+
+  void AddKeyAndAlt(const Slice& key, const Slice& alt) override {
+    uint64_t key_hash = GetSliceHash64(key);
+    uint64_t alt_hash = GetSliceHash64(alt);
+    std::optional<uint64_t> prev_key_hash;
+    std::optional<uint64_t> prev_alt_hash = hash_entries_info_.prev_alt_hash;
+    if (!hash_entries_info_.entries.empty()) {
+      prev_key_hash = hash_entries_info_.entries.back();
+    }
+    // Add alt first, so that entries.back() always contains previous key
+    // ASSUMING a change from one alt to the next implies a change to
+    // corresponding key
+    if (alt_hash != prev_alt_hash && alt_hash != key_hash &&
+        alt_hash != prev_key_hash) {
+      AddHash(alt_hash);
+    }
+    // Overwrite prev_alt_hash for cases like alt_hash == key_hash
```

Comment:
nit: this overwrites prev_alt_hash regardless of whether alt_hash and key_hash is the same, right?

Reply:
Yes. It's possible for a prefix to equal the key it comes from. E.g.

foo,foo // <- add "foo" as key but also as prev_alt
foo123,foo // <- "foo" is still prev_alt
foo456,foo // <- depends on prev_alt having been updated despite earlier matching of prev_key or key